src_analysis/fuzzer_dependencies.py

```diff
@@ -0,0 +1,182 @@
+    fuzzer_directory = fuzzer_utils.FuzzerDirectory(fuzzer)
+
+    if initial_fuzzer == fuzzer:
+        dependencies = set()
+    else:
+        # If fuzzer's base fuzzer is different, fuzzer is a variant, which
+        # means changes to variants.yaml can affect it.
+        dependencies = {fuzzer_directory.variants_yaml}
+
+    # Don't use modulefinder for python dependencies since it doesn't work
+    # without __init__.py files.
+    fuzzer_module = importlib.import_module(_get_fuzzer_module_name(fuzzer))
+    dependencies = dependencies.union(_get_python_dependencies(fuzzer_module))
+
+    # The fuzzer is also dependent on dockerfiles.
+    dependencies = dependencies.union(fuzzer_directory.dockerfiles)
+    return dependencies
+
+
+def _get_python_dependencies(module: types.ModuleType,
+                             depth: int = 0) -> Set[str]:
+    """Returns the python files that |module| is dependent on if module is a
+    submodule of fuzzers/. Does not return dependencies that are not submodules
+    of fuzzers/, such as std library modules. Has a limit of
+    PY_DEPENDENCIES_MAX_DEPTH so that cyclic imports can easily be detected.
+    Note that this may not work if a fuzzer.py imports modules dynamically or
+    within individual functions. That is ok because we can prevent this during
+    code review."""
+    if depth > PY_DEPENDENCIES_MAX_DEPTH:
+        # Enforce a depth to catch cyclic imports.
+        format_string = ('Depth: {} greater than max: {}. '
+                         'Probably a cyclic import in {}.')
+        raise Exception(
+            format_string.format(depth, PY_DEPENDENCIES_MAX_DEPTH, module))
+
+    module_path = inspect.getfile(module)
+
+    # Just get the dependencies from the cache if we have them.
+    # This would break on modules doing crazy things like writing Python files
+    # and then importing them, code review should prevent that from landing
+    # though.
+    if module_path in PY_DEPENDENCIES_CACHE:
+        return PY_DEPENDENCIES_CACHE[module_path]
+
+    # Modules depend on themselves.
+    dependencies = {module_path}
+
+    # This assumes that every module that |module| depends on is imported in
+    # top-level code.
+    attr_names = dir(module)
+    for attr_name in attr_names:
+        # Check if attr_name is a module, bail if not.
+        imported_module = getattr(module, attr_name)
+        if not isinstance(imported_module, types.ModuleType):
+            continue
+
+        # Files in fuzzers/ can only import code from the Python standard
+        # library modules and modules in fuzzers/.
+        if is_fuzzers_submodule(imported_module):
+            imported_module_path = inspect.getfile(imported_module)
+            dependencies.add(imported_module_path)
+            # Now recur to get the dependencies of the dependency.
+            dependencies = dependencies.union(
+                _get_python_dependencies(imported_module, depth + 1))
+
```

Comment:
global PY_DEPENDENCIES_CACHE ?

Reply:
I don't think this is needed since we are mutating the `PY_DEPENDENCIES_CACHE`. If we were reassigning where we would need to do this.